client/fetch_state.go

```diff
@@ -0,0 +1,169 @@
+
+package client
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/encoding"
+	"github.com/m3db/m3db/serialize"
+	"github.com/m3db/m3db/storage/index"
+	"github.com/m3db/m3db/topology"
+	"github.com/m3db/m3db/x/xpool"
+	"github.com/m3db/m3x/ident"
+)
+
+const (
+	maxUint = ^uint(0)
+	maxInt  = int(maxUint >> 1)
+)
+
+var (
+	errFetchStateStillProcessing = errors.New("[invariant violated] fetch state is still processing, unable to create response")
+)
+
+type fetchState struct {
+	sync.Cond
+	sync.Mutex
+	refCounter
+
+	op                   *fetchTaggedOp
+	tagResultAccumulator fetchResultsAccumulator
+	err                  error
+	done                 bool
+
+	pool fetchStatePool
+}
+
+func newFetchState(pool fetchStatePool) *fetchState {
+	f := &fetchState{
+		tagResultAccumulator: &fetchTaggedResultAccumulator{},
+		pool:                 pool,
+	}
+	f.destructorFn = f.close // Set refCounter completion as close
+	f.L = f                  // Set the embedded condition locker to the embedded mutex
+	return f
+}
+
+func (f *fetchState) close() {
+	if f.op != nil {
+		f.op.decRef()
+		f.op = nil
+	}
+	f.err = nil
+	f.done = false
+	f.tagResultAccumulator.Clear()
+
+	if f.pool == nil {
+		return
+	}
+	f.pool.Put(f)
+}
+
+func (f *fetchState) Reset(
+	startTime time.Time,
+	endTime time.Time,
+	op *fetchTaggedOp,
+	topoMap topology.Map,
+	majority int,
+	consistencyLevel topology.ReadConsistencyLevel,
+) {
+	op.incRef() // take a reference to the provided op
+	f.op = op
+	f.tagResultAccumulator.Reset(startTime, endTime, topoMap, majority, consistencyLevel)
+}
+
+func (f *fetchState) completionFn(
+	result interface{},
+	resultErr error,
+) {
+	f.Lock()
+	defer func() {
+		f.Unlock()
+		f.decRef() // release ref held onto by the hostQueue (via op.completionFn)
+	}()
+
+	if f.done {
+		// i.e. we've already failed, no need to continue processing any additional
+		// responses we receive
+		return
+	}
+
+	done, err := f.tagResultAccumulator.Add(result, resultErr)
+	if done {
+		f.err = err
+		f.done = true
+		f.Signal()
+	}
+}
+
+func (f *fetchState) asIndexQueryResults(pools fetchTaggedPools) (index.QueryResults, error) {
+	f.Lock()
+	defer f.Unlock()
+
+	if !f.done {
+		return index.QueryResults{}, errFetchStateStillProcessing
+	}
+
+	if err := f.err; err != nil {
+		return index.QueryResults{}, err
+	}
+
+	limit := maxInt
+	if l := f.op.request.Limit; l != nil {
+		limit = int(*l)
+	}
+
+	return f.tagResultAccumulator.AsIndexQueryResults(limit, pools)
+}
+
+func (f *fetchState) asEncodingSeriesIterators(pools fetchTaggedPools) (encoding.SeriesIterators, bool, error) {
+	f.Lock()
+	defer f.Unlock()
+
+	if !f.done {
+		return nil, false, errFetchStateStillProcessing
+	}
+
+	if err := f.err; err != nil {
+		return nil, false, err
+	}
+
+	limit := maxInt
+	if l := f.op.request.Limit; l != nil {
+		limit = int(*l)
+	}
+
+	return f.tagResultAccumulator.AsEncodingSeriesIterators(limit, pools)
+}
+
+type fetchTaggedPools interface {
```

Comment:
Only have it as an interface to override with a stub implementation of the pool in tests (see `fetch_state_test.go`). Won't the runtime optimize away the itab considering there's only one backing implementation?

Reply:
Yeah I'm not worried about the optimization.  Guess I don't see how different it is though using an interface or struct for testing purposes too since you can just as easily override the struct field in a test vs using `pools.EXPECT().SomeAccessor().Return(...)`?

Reply:
Ah sorry thought you were talking about another pool. Put a blurb explaining what we talked about offline.